tools/comdb2ar/increment.cpp

```diff
@@ -0,0 +1,523 @@
+
+    for(std::vector<uint32_t>::iterator
+            it = pages.begin();
+            it != pages.end();
+            ++it){
+
+        ifs.seekg(pagesize * *it, ifs.beg);
+        ifs.read(&char_pagebuf[0], pagesize);
+        ssize_t bytes_read = ifs.gcount();
+        if(bytes_read < pagesize) {
+            std::ostringstream ss;
+            ss << "read error";
+            throw SerialiseError(filename, ss.str());
+        }
+
+        std::memcpy(pagebuf, char_pagebuf, pagesize);
+
+        uint32_t cksum;
+        bool cksum_verified;
+        verify_checksum(pagebuf, pagesize, false, false,
+                            &cksum_verified, &cksum);
+        if(!cksum_verified){
+            std::ostringstream ss;
+            ss << "serialise_file:page failed checksum verification";
+            throw SerialiseError(filename, ss.str());
+        }
+
+        // Update the diff .incr file
+        incrFile.seekp(12 * *it, incrFile.beg);
+        PAGE * pagep = (PAGE *) pagebuf;
+
+        incrFile.write((char *) &(LSN(pagep).file), 4);
+        incrFile.write((char *) &(LSN(pagep).offset), 4);
+        incrFile.write((char *) &cksum, 4);
+
+        ssize_t bytes_written = writeall(1, pagebuf, pagesize);
+        if(bytes_written != pagesize){
+            std::ostringstream ss;
+            ss << "error writing text data: " << std::strerror(errno);
+            throw SerialiseError(filename, ss.str());
+        }
+
+        total_read += bytes_read;
+        total_written += bytes_written;
+        if(total_written != total_read){
+            std::ostringstream ss;
+            ss << "different amounts written and read. wrote " << total_written
+                << " bytes and read " << total_read << " bytes "
+                << std::strerror(errno);
+            throw SerialiseError(filename, ss.str());
+        }
+    }
+
+    if (pagebuf)
+        free(pagebuf);
+
+    std::clog << "a " << filename << " pages=[";
+    for(size_t i = 0; i < pages.size(); ++i){
+        std::clog << pages[i];
+        if(i != pages.size() - 1){
+            std::clog << " ";
+        } else {
+            std::clog << "] ";
+        }
+    }
+    std::clog << "pagesize=" << pagesize << std::endl;
+
+    return total_read;
+}
+
+// Get the datetime as a string YYYYMMDDHHMMSS
+std::string getDTString() {
+    time_t rawtime;
+    struct tm *timeinfo;
+    char buffer[80];
+
+    time(&rawtime);
+    timeinfo = localtime(&rawtime);
+
+    strftime(buffer, sizeof(buffer), "%Y%m%d%I%M%S", timeinfo);
+    std::string str(buffer);
+
+    return str;
+}
+
+void incr_deserialise_database(
+    const std::string& lrldestdir,
+    const std::string& datadestdir,
+    const std::string& dbname,
+    std::set<std::string>& table_set,
+    std::string& sha_fingerprint,
+    unsigned percent_full,
+    bool force_mode,
+    bool& is_disk_full,
+    const std::string& incr_path,
+    bool keep_all_logs
```

Comment:
I think we discussed this one - we should never keep all logs from all the increments: full-recovery will never work if there are holes in the logstream.  Keep only the logfiles from the last increment.